streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -306,12 +307,15 @@ private void restoreState(final StateRestoreCallback stateRestoreCallback,
     @Override
     public void flush() {
         log.debug("Flushing all global globalStores registered in the state manager");
-        for (final StateStore store : this.globalStores.values()) {
-            try {
-                log.trace("Flushing global store={}", store.name());
-                store.flush();
-            } catch (final Exception e) {
-                throw new ProcessorStateException(String.format("Failed to flush global state store %s", store.name()), e);
+        for (final Optional<StateStore> maybeStore : globalStores.values()) {
```

Comment:
After the task has been transited to `running`, it should mean that all its state stores have been successfully initialized and restored, so this should not happen. If it ever happens we should treat it as IllegalState indicating Streams bug. 


Ditto elsewhere.

Reply:
Actually thinking about it: today we can reinitialize in three cases:

1) for active / standby / global task, when offset out of range during `restoring`.
2) for standby task, when offset out of range during `running`.
3) for active / standby / global task, when checkpoint not found and EOS enabled when starting `restoring`.

I.e. it seems to me that for any `task#flush / checkpoint` call should have all `present` stores, while for `task#close` some stores may not be `present` if an exception in the store#initialize caused the whole task to be closed. So I suggest we throw Illegal for the former functions, while logging an INFO for the latter.

Reply:
I'll implement your suggestions when I get a chance.

I can see the high-level value of tightening our assertions about what state these things can be in, to catch more bugs. But on the other hand, does it make sense for the StateManager to be that opinionated about the lifecycle of the Task that contains it?

Reply:
One note, the code I've submitted is functionally equivalent to the pre-existing code, since under the situation where a `maybeStore` is `empty`, in the existing codebase, the mapping would simply be absent from the `stores` map and therefore skipped in this iteration.

Reply:
Yeah that's true. I was trying to tighten the screw a bit in case we have some lurking bugs, it would be easier to detect.

That being said, I think I also agree having opinionated StateManager depending on other classes is not a good design pattern. So how about taking one step back from aggressiveness and just log it when we are skipping on these functions?

Reply:
Cool. I didn't get a chance to look at it yet, but I'll see how I can best accomplish the goal.

Reply:
Ok, I looked around, and I agree it should be fine to put in those else conditions. Let's see if the tests pass! ;)